src/builders.ts

```diff
@@ -1,18 +1,122 @@
 import { isAbsolute, relative } from 'path';
+import { EOL } from 'os';
 import slash = require('slash');
 import { createHash } from 'crypto';
 import {
   DaysToDecay,
   GenericLintData,
-  Location,
+  Operation,
   TodoConfig,
-  TodoData,
-  TodoDataV1,
   TodoDataV2,
   TodoDates,
   TodoFileFormat,
+  TodoFileHash,
+  TodoFilePathHash,
 } from './types';
 import { getDatePart } from './date-utils';
+import { todoDirFor } from './io';
+import TodoMatcher from './todo-matcher';
+
+const SEPARATOR = '|';
+
+export function buildFromTodoOperations(
+  todoOperations: string[]
+): Map<TodoFilePathHash, TodoMatcher> {
+  const existingTodos = new Map<TodoFilePathHash, TodoMatcher>();
+
+  for (const todoOperation of todoOperations) {
+    const [
+      operation,
+      engine,
+      ruleId,
+      line,
+      column,
+      endLine,
+      endColumn,
+      source,
+      fileFormat,
+      createdDate,
+      warnDate,
+      errorDate,
+      ...filePathSegments
+    ] = todoOperation.split(SEPARATOR);
+
+    // The only case where we need to join back on the separator is when the filePath itself
+    // contains a pipe ('|') char. Normal filePaths will simply join without the separator.
+    const filePath = filePathSegments.join(SEPARATOR);
+    const todoFileDir = todoDirFor(filePath);
+
+    if (!existingTodos.has(todoFileDir)) {
+      existingTodos.set(todoFileDir, new TodoMatcher());
+    }
+
+    const matcher = existingTodos.get(todoFileDir);
+
+    matcher?.addOrRemove(<Operation>operation, {
+      engine,
+      ruleId,
+      filePath,
+      fileFormat: Number.parseInt(fileFormat, 10),
+      range: {
+        start: {
+          line: Number.parseInt(line, 10),
+          column: Number.parseInt(column, 10),
+        },
+        end: {
+          line: Number.parseInt(endLine, 10),
+          column: Number.parseInt(endColumn, 10),
+        },
+      },
+      source,
+      createdDate: Number.parseInt(createdDate, 10),
+      warnDate: Number.parseInt(warnDate, 10),
+      errorDate: Number.parseInt(errorDate, 10),
+    });
+  }
+
+  for (const [filePath, matcher] of existingTodos.entries()) {
+    if (matcher.unprocessed.size === 0) {
+      existingTodos.delete(filePath);
+    }
+  }
+
+  return existingTodos;
+}
+
+export function buildTodoOperations(
+  add: Map<TodoFileHash, TodoDataV2>,
+  remove: Map<TodoFileHash, TodoDataV2>
+): string {
+  const ops: string[] = [];
+
+  for (const [, todoDatum] of add) {
+    ops.push(toOperation('add', todoDatum));
+  }
+
+  for (const [, todoDatum] of remove) {
+    ops.push(toOperation('remove', todoDatum));
+  }
+
+  return ops.join(EOL) + EOL;
```

Comment:
Can you test calling this with `buildTodoOperations(new Set(), new Set())`? I think we'll end up with an extra trailing newline.

Reply:
Fixed and added tests.